ambry-store/src/main/java/com.github.ambry.store/BlobStore.java

```diff
@@ -167,27 +166,27 @@ public void put(MessageWriteSet messageSetToWrite) throws StoreException {
       synchronized (lock) {
         // Validate that log end offset was not changed. If changed, check once again for existing
         // keys in store
-        long currentIndexEndOffset = index.getCurrentEndOffset();
-        if (currentIndexEndOffset != indexEndOffsetBeforeCheck) {
+        Offset currentIndexEndOffset = index.getCurrentEndOffset();
+        if (!currentIndexEndOffset.equals(indexEndOffsetBeforeCheck)) {
           FileSpan fileSpan = new FileSpan(indexEndOffsetBeforeCheck, currentIndexEndOffset);
           for (MessageInfo info : messageSetToWrite.getMessageSetInfo()) {
             if (index.findKey(info.getStoreKey(), fileSpan) != null) {
               throw new StoreException("Key already exists on filespan check", StoreErrorCodes.Already_Exist);
             }
           }
         }
-        // TODO (Index changes): Working under the assumption that there is only one log segment.
-        long currentLogEndOffset = log.getEndOffset().getOffset();
+        Offset currentLogEndOffset = log.getEndOffset();
         messageSetToWrite.writeTo(log);
```

Comment:
I agree with you and it makes me uncomfortable too. But unless the `writeSet` has a way of communicating this, I don't think we can make it any better.

I agree with having it one class and will move `getFileSpanForMessage()` to the `Log`. 

Regarding the interface change you suggest, it can be explored but since it involves touching interfaces outside the store package - if we are going to do it, we should do it another PR.  Also, I'm not sure what the interface change would buy us. Generating one by one as they are written v/s generating after everything is written should be no different in terms of complexity/bugs etc right?